### PR TITLE
Interrupt and sigterm before sigkill kernel.

### PR DESCRIPTION
**Pull request by Carreau**
This implements the discussion in #618, to try to more progressively stop
the kernels.

1) this always sends and interrupt before any shutdown requests;
  goal being to stop any processing happening that may block any event
  loop.

2) sends the shutdown_requests (no changes here).

3) wait 50% of the wait time, and sends a "terminate" in quote as this
depends on your platform/os, and the type of kernel you have.
    - if subprocess.Popen calls `.terminate()` which is SIGTERM on unix; the same as
    `.kill()` on windows; if not a Popen instance send SIGTERM.

4) wait the other 50% and kills, (same as before).

This does this both on the sync client and async client.

TBD: write tests and docs; and test more properly;

### CODE REVIEW

**Comment by Carreau** [comment]
I'm thinking about changing the `shutdown_kernel` to return an object with how far it went to stop the kernel; something that we can grow later; which basically tell us if the kernel replied to the shutdown_request; or had to go to the sigterm, or the sigkill. That might make it easier to test (as it is hard to get info from a dead process), and thus implement multiple kernels that will forcibly not shutdown on request, and not shut down on sigterm.

**Review by MSeal — APPROVED** [review]
Overall looks fine beyond some minor comments. Might be worth making a manual test plan for this one to make sure we have someone try it out in a few event loops patterns / os's? Just because I could see some odd edge case in windows for M1 osx that our automation might not catch.

File: jupyter_client/manager.py [hunk 1]
```diff
@@ -29,13 +31,23 @@
     KernelManagerABC
 )
 
+class _ShutdownStatus(Enum):
+    Unset = None
+    ShutdownRequest = "ShutdownRequest"
+    SigtermRequest = "SigtermRequest"
+    SigKillRequest = "SigKillRequest"
```
Comment (kevin-bates) [inline jupyter_client/manager.py:38]: nit: Could you please either lowercase the 'K' in `SigKillRequest` or raise the 'T' in `SigtermRequest`?  I'd prefer the former.

File: jupyter_client/manager.py [hunk 2]
```diff
@@ -584,9 +647,23 @@ async def finish_shutdown(self, waittime=None, pollinterval=0.1):
...
+                    self._async_wait(pollinterval=pollinterval), timeout=waittime / 2
+                )
+            except asyncio.TimeoutError:
+                self.log.debug("Kernel is taking too long to finish, terminating")
+                self._shutdown_status = _ShutdownStatus.SigtermRequest
+                await self._send_kernel_sigterm()
+
+            await asyncio.wait_for(
+                self._async_wait(pollinterval=pollinterval), timeout=waittime / 2
+            )
```
Comment (kevin-bates) [inline jupyter_client/manager.py:662]: I think this spin-wait needs to be associated with the `except` block - otherwise, it will occur again for the successful shutdown-request case (which will immediately exit, but probably not the intent).

File: jupyter_client/manager.py [hunk 3]
```diff
@@ -584,9 +647,23 @@ async def finish_shutdown(self, waittime=None, pollinterval=0.1):
...
             waittime = max(self.shutdown_wait_time, 0)
         try:
-            await asyncio.wait_for(self._async_wait(pollinterval=pollinterval), timeout=waittime)
+            try:
+                self._shutdown_status = _ShutdownStatus.ShutdownRequest
+                await asyncio.wait_for(
+                    self._async_wait(pollinterval=pollinterval), timeout=waittime / 2
+                )
+            except asyncio.TimeoutError:
+                self.log.debug("Kernel is taking too long to finish, terminating")
```
Comment (kevin-bates) [inline jupyter_client/manager.py:656]: We should add this debug statement to the synchronous version.

**Review by kevin-bates — COMMENTED** [review]
(Sorry, I fat-fingered the 'single-comment' button, and its like it turned previous comments into single comments as well, so I continued in that mode.)

This is looking good and I believe introducing interrupt and terminate to the shutdown mix is great.

I don't think this is anything to change, but just noting that the shutdown-request status is set after the request, while the signal-based statuses are set prior to their respective requests.  Although not consistent, I think it's probably the right approach across the three and I'm not advocating for additional (pre/post) statuses, but with Kernel Provisioning, the `terminate()` and `kill()` are not necessarily "signal" requests and we shouldn't assume that a shtudown_status of `SigtermRequest` has been completely delivered.  (Assuming there were a means of checking the status - which I could see some admin-level apps wanting to do.)

Hmm - on that note, should we consider name changes to `TerminateRequest` and `KillRequest`.  These sound like kernel messages.  I suppose we could use `Request` as a _prefix_ on all three: `RequestShutdown`, `RequestTerminate`, and `RequestKill`.

File: jupyter_client/manager.py [hunk 4]
```diff
@@ -340,10 +361,25 @@ def finish_shutdown(self, waittime=None, pollinterval=0.1):
...
+            # send sigterm and wait the other 50%.
+            self._shutdown_status = _ShutdownStatus.SigtermRequest
+            self._send_kernel_sigterm()
+            for i in range(int(waittime / 2 / pollinterval)):
+                if self.is_alive():
+                    time.sleep(pollinterval)
+                else:
+                    # If there's still a proc, wait and clear
+                    if self.has_kernel:
+                        self.kernel.wait()
```
[the remark this answers is not in the document]
Comment (Carreau) [reply jupyter_client/manager.py:374]: I don't think so, that's the same logic as  line 360; I could make it a local function and call it in both location to be clearer.

File: jupyter_client/manager.py (hunk 2 above)
Comment (Carreau) [reply jupyter_client/manager.py:662]: Yeah, it's with the one just below, let me try to change the layout to be clearer.

File: jupyter_client/manager.py (hunk 1 above)
Comment (Carreau) [reply jupyter_client/manager.py]: Done.

File: jupyter_client/manager.py [hunk 5]
```diff
@@ -458,6 +496,31 @@ def has_kernel(self):
...
+        """similar to _kill_kernel, but with sigterm (not sigkill), but do not block"""
+        if self.has_kernel:
+            # Signal the kernel to terminate (sends SIGTERM on Unix and
+            # if the kernel is a subprocess and we are on windows; this is
+            # equivalent to kill
+            try:
+                if hasattr(self.kernel, "terminate"):
+                    self.kernel.terminate()
+                elif hasattr(signal, "SIGTERM"):
+                    self.signal_kernel(signal.SIGTERM)
```
[the remark this answers is not in the document]
Comment (Carreau) [reply jupyter_client/manager.py:516]: added an else with a log statement; I'm not too sure of cases where we can't send sigterm; 
That would likely be windows if the kernel is not a subprocess; in which case we won't send anything, but will send sigkill equivalent anyway after the full wait time.

File: jupyter_client/manager.py (hunk 2 above)
Comment (Carreau) [reply jupyter_client/manager.py:665]: I've split the two try/except to make the intention clearer, is that better ?

File: jupyter_client/manager.py [hunk 6]
```diff
@@ -673,6 +752,31 @@ async def restart_kernel(self, now=False, newports=False, **kw):
...
+        """similar to _kill_kernel, but with sigterm (not sigkill), but do not block"""
+        if self.has_kernel:
+            # Signal the kernel to terminate (sends SIGTERM on Unix and
+            # if the kernel is a subprocess and we are on windows; this is
+            # equivalent to kill
+            try:
+                if hasattr(self.kernel, "terminate"):
+                    self.kernel.terminate()
+                elif hasattr(signal, "SIGTERM"):
+                    await self.signal_kernel(signal.SIGTERM)
```
[the remark this answers is not in the document]
Comment (Carreau) [reply jupyter_client/manager.py:776]: done

File: jupyter_client/tests/signalkernel.py [hunk 7]
```diff
@@ -25,6 +29,14 @@ def __init__(self, **kwargs):
         super().__init__(**kwargs)
         self.children = []
 
+        if os.environ.get("NO_SIGTERM_REPLY", None) == "1":
+            signal.signal(signal.SIGTERM, signal.SIG_IGN)
+
+    @gen.coroutine
+    def shutdown_request(self, stream, ident, parent):
+        if os.environ.get("NO_SHUTDOWN_REPLY") != "1":
+            yield gen.maybe_future(super().shutdown_request(stream, ident, parent))
```
[the remark this answers is not in the document]
Comment (Carreau) [reply jupyter_client/tests/signalkernel.py:38]: I don't believe so,  that's the exact way ipykernel uses coroutines:

```
 @gen.coroutine
    def shutdown_request(self, stream, ident, parent):
        content = yield gen.maybe_future(self.do_shutdown(parent['content']['restart']))
        ...
```

So if this breaks; IPykernel would be broken, and if it works with ipykernel, I see no reason this wouldn't. 

I would prefer `async def` and `await`, but that unfortunately does not mix.

File: jupyter_client/manager.py [hunk 8]
```diff
@@ -507,6 +514,11 @@ def _send_kernel_sigterm(self):
                     self.kernel.terminate()
                 elif hasattr(signal, "SIGTERM"):
                     self.signal_kernel(signal.SIGTERM)
+                else:
+                    self.log.debug(
+                        "Cannot set term signal to kernel, no"
```
Comment (kevin-bates) [inline jupyter_client/manager.py:519]: I see we prefer `SIGKILL` over `kill()` and assume that `kill()` exists.  Should we do the same with `SIGTERM` and `terminate()`? (I suspect this is due to Windows not having `SIGTERM`?) 

Actually, given that `Popen` is well-documented to use `SIGKILL` and `SIGTERM` for `kill()` and `terminate()`, respectively, and there don't appear to be python version compatibility issues, I think we'd be fine just calling the desired methods directly.  This would then forgo the need to log anything and remove the conditional attribute checks.

**Review by kevin-bates — COMMENTED** [review]
Thanks for the update Matthias - they look good.

I just had the one comment regarding possibly cutting to the chase and using the subprocess methods unconditionally.  There might be some history here that I'm not seeing.  Thanks.

**Comment by Carreau** [comment]
> I don't think this is anything to change, but just noting that the shutdown-request status is set after the request, while the signal-based statuses are set prior to their respective requests. Although not consistent, I think it's probably the right approach across the three and I'm not advocating for additional (pre/post) statuses, but with Kernel Provisioning, the `terminate()` and `kill()` are not necessarily "signal" requests and we shouldn't assume that a shtudown_status of `SigtermRequest` has been completely delivered. (Assuming there were a means of checking the status - which I could see some admin-level apps wanting to do.)
> 
> Hmm - on that note, should we consider name changes to `TerminateRequest` and `KillRequest`. These sound like kernel messages. I suppose we could use `Request` as a _prefix_ on all three: `RequestShutdown`, `RequestTerminate`, and `RequestKill`.

I do not have any preference, this is mostly for testing in order to understand the internal state and why I kept them private. ; I initially thought of returning the values but I'm not feeling confident with making these public.

File: jupyter_client/manager.py (hunk 8 above)
Comment (Carreau) [reply jupyter_client/manager.py:526]: Diggning in:

I don't think that's compatibility issue, signal_kernel signal the process group, not the kernel alone.
Sigterm can be handled by a subprocess, not sigkill.

Signal_kernel send the signal to the process group (when possible) instead of only the single process, so if the kernel has children you don't want to send sigterm to the (grand)children, put leave the child process a chance to clean up it children, in case there is some cleanup logic.

Thus `terminate()` is prefered, and if you can't you sigterm  the process group. 

Kill is the opposite, as the kernel has no chance to terminate children by catching sigkill, you want to tell them to die  as well. See #314 

I guess in a perfect world you would
 - 1 shutdown request
 - 2 wait
 - 3 term child
 - 4 wait
    -  4b if don't respond term grand children
 - 5 wait
     - 5b kill everybody. 

we just skip 4,4b, and replace 3 by "term child, if you can't term everybody".

Comment (kevin-bates) [reply jupyter_client/manager.py:526]: Good point on the process group portion of things.  I had conflated `signal_kernel()` with `send_signal()`.

Comment (Carreau) [reply jupyter_client/manager.py:526]: Yeah, i I had to dig deep in the code and history to figure that one :-)

**Review by kevin-bates — APPROVED** [review]
Thank you Matthias (and @mlucool for proposing these changes) - this makes a kernel's management more robust.

**Comment by mlucool** [comment]
Thanks @Carreau - I tested this and it does fix https://github.com/jupyter/jupyter_client/issues/618

**Comment by Carreau** [comment]
there has been no objecting in a week; merging. Will do a release soon.
